src/main/java/org/opensearch/neuralsearch/util/NeuralQueryValidationUtil.java

```diff
@@ -0,0 +1,274 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.opensearch.neuralsearch.util;
+
+import lombok.NonNull;
+import org.apache.commons.lang.StringUtils;
+import org.opensearch.index.mapper.RankFeaturesFieldMapper;
+import org.opensearch.knn.index.mapper.KNNVectorFieldMapper;
+import org.opensearch.neuralsearch.query.NeuralQueryBuilder;
+import org.opensearch.neuralsearch.query.dto.NeuralQueryBuildStage;
+import org.opensearch.neuralsearch.query.dto.NeuralQueryTargetFieldConfig;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+
+import static org.opensearch.knn.index.query.KNNQueryBuilder.EXPAND_NESTED_FIELD;
+import static org.opensearch.knn.index.query.KNNQueryBuilder.FILTER_FIELD;
+import static org.opensearch.knn.index.query.KNNQueryBuilder.K_FIELD;
+import static org.opensearch.knn.index.query.KNNQueryBuilder.MAX_DISTANCE_FIELD;
+import static org.opensearch.knn.index.query.KNNQueryBuilder.METHOD_PARAMS_FIELD;
+import static org.opensearch.knn.index.query.KNNQueryBuilder.MIN_SCORE_FIELD;
+import static org.opensearch.knn.index.query.KNNQueryBuilder.RESCORE_FIELD;
+import static org.opensearch.neuralsearch.common.MinClusterVersionUtil.isClusterOnOrAfterMinReqVersionForDefaultDenseModelIdSupport;
+import static org.opensearch.neuralsearch.query.NeuralQueryBuilder.MODEL_ID_FIELD;
+import static org.opensearch.neuralsearch.query.NeuralQueryBuilder.QUERY_IMAGE_FIELD;
+import static org.opensearch.neuralsearch.query.NeuralQueryBuilder.QUERY_TEXT_FIELD;
+import static org.opensearch.neuralsearch.query.NeuralQueryBuilder.QUERY_TOKENS_FIELD;
+
+public class NeuralQueryValidationUtil {
+    public static List<String> validateNeuralQueryForKnn(
+        @NonNull final NeuralQueryBuilder queryBuilder,
+        final NeuralQueryBuildStage buildStage
+    ) {
+        final List<String> errors = validateNeuralQueryForDenseCommonRules(queryBuilder);
+
+        if (isModelIdRequiredForNeuralQueryForKnn(buildStage)) {
+            if (queryBuilder.modelId() == null) {
+                errors.add(String.format(Locale.ROOT, "%s must be provided.", MODEL_ID_FIELD.getPreferredName()));
+            }
+        }
+        return errors;
+    }
+
+    private static boolean isModelIdRequiredForNeuralQueryForKnn(final NeuralQueryBuildStage buildStage) {
+        // If we build the NeuralQueryBuilder targeting a knn field in the coordinate rewrite then it must have a model id.
+        // If we build the NeuralQueryBuilder targeting a knn field in the fromXContent function then it must have a
+        // model id if it cannot support the default model id otherwise the model id is optional.
+        return NeuralQueryBuildStage.REWRITE.equals(buildStage)
+            || ((buildStage == null || NeuralQueryBuildStage.FROM_X_CONTENT.equals(buildStage))
+                && isClusterOnOrAfterMinReqVersionForDefaultDenseModelIdSupport() == false);
```

Comment:
The nested conditions are too complex for readability, maybe like:

```suggestion
    // During rewrite stage, model ID is always required
    if (NeuralQueryBuildStage.REWRITE.equals(buildStage)) {
        return true;
    }
    
    // During FROM_X_CONTENT or unspecified stage, model ID is required if default model ID isn't supported
    boolean isFreshBuild = (buildStage == null || NeuralQueryBuildStage.FROM_X_CONTENT.equals(buildStage));
    boolean supportsDefaultModel = isClusterOnOrAfterMinReqVersionForDefaultDenseModelIdSupport();
    
    return isFreshBuild && !supportsDefaultModel;
```

Reply:
ok